operator/controllers/reconcilers/seldon/runtime_reconciler.go

```diff
@@ -40,80 +49,124 @@ func ParseInt32(s string, defaultVal int32) (int32, error) {
 
-func ValidateDataflowScaleSpec(
+func getEnvVarValue(podSpec *v1.PodSpec, name string, defaultValue string) string {
+	if podSpec != nil && len(podSpec.Containers) > 0 {
+		for _, env := range podSpec.Containers[0].Env {
+			if env.Name == name {
+				return env.Value
+			}
+		}
+	}
+
+	return defaultValue
+}
+
+func min(a, b int32) int32 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func replicaCalc(resourceCount, maxConsumers, partitions int32) int32 {
+	if resourceCount == 0 {
+		return 1
+	}
+	return partitions * min(resourceCount, maxConsumers)
+}
+
+func validateScaleSpec(
 	component *mlopsv1alpha1.ComponentDefn,
 	runtime *mlopsv1alpha1.SeldonRuntime,
 	commonConfig common.ReconcilerConfig,
 	namespace *string,
+	envVar string,
+	defaultMaxConsumers int32,
+	eventReason string,
+	resourceListObj client.ObjectList,
+	countResources func(client.ObjectList) int,
 ) error {
 	ctx, clt, recorder := commonConfig.Ctx, commonConfig.Client, commonConfig.Recorder
-	numPartitions, err := ParseInt32(runtime.Spec.Config.KafkaConfig.Topics["numPartitions"].StrVal, 1)
+
+	numPartitions, err := ParseInt32(
+		runtime.Spec.Config.KafkaConfig.Topics["numPartitions"].StrVal,
+		DEFAULT_NUM_PARTITIONS,
+	)
 	if err != nil {
 		return fmt.Errorf("failed to parse numPartitions from KafkaConfig: %w", err)
 	}
 
-	var pipelineCount int32 = 0
+	var resourceCount int32 = 0
 	if namespace != nil {
-		// Get the number of Pipeline resources in the namespace
-		var pipelineList mlopsv1alpha1.PipelineList
-		if err := clt.List(ctx, &pipelineList, client.InNamespace(*namespace)); err != nil {
-			return fmt.Errorf("failed to list Pipeline resources in namespace %s: %w", *namespace, err)
+		if err := clt.List(ctx, resourceListObj, client.InNamespace(*namespace)); err != nil {
+			return fmt.Errorf("failed to list resources in namespace %s: %w", *namespace, err)
 		}
-		pipelineCount = int32(len(pipelineList.Items))
+		resourceCount = int32(countResources(resourceListObj))
 	}
 
-	maxReplicas := numPartitions
-	if pipelineCount != 0 {
-		maxReplicas = numPartitions * pipelineCount
+	var maxConsumers int32 = defaultMaxConsumers
+	if envVar != "" {
+		envValue := getEnvVarValue(component.PodSpec, envVar, "")
+		maxConsumers, err = ParseInt32(envValue, defaultMaxConsumers)
+		if err != nil {
+			return fmt.Errorf("failed to parse %s: %w", envVar, err)
+		}
+		if maxConsumers == 0 {
+			return fmt.Errorf("invalid %s value: %s", envVar, envValue)
+		}
 	}
 
+	maxReplicas := replicaCalc(resourceCount, maxConsumers, numPartitions)
 	if component.Replicas != nil && *component.Replicas > maxReplicas {
 		component.Replicas = &maxReplicas
 		recorder.Eventf(
 			runtime,
 			v1.EventTypeWarning,
-			"DataflowEngineReplicasAdjusted",
-			fmt.Sprintf("Dataflow engine replicas adjusted to %d based on KafkaConfig and Pipeline count", maxReplicas),
+			eventReason,
+			fmt.Sprintf("%s replicas adjusted to %d based on KafkaConfig and resource count", component.Name, maxReplicas),
```

Comment:
The message should reflect that the number of requested replicas was above the maximum possible for the given configuration, and only then that it was adjusted.